### DISCUSSION TITLE
Quorum queue message order issue when consumers disconnect

### DISCUSSION THREAD

**Discussion by anhanhnguyen** [rabbitmq/rabbitmq-server#6805] [Other]
Hi,

We found an unexpected behaviour with quorum queues when doing get messages, nack and requeue.

The testing procedure:
1. Consume messages from a quorum queue with a prefect count = 5. Do not acknowledge the messages
2. Kill the consumer (drop the connection)
3. Get messages from UI with nack and requeue true

When we get the messages from the UI, we expect to get the first message in the queue and it should be the same message every time. However, we get different messages each time.

Example:
Publish messages 1, 2, 3, 4, 5, 6, 7, 8. 
Consume with prefetch = 5, do not acknowledge. 
Kill the consumer. 

Get a message from the queue. 
Returns message 1.
Get a message from the queue. 
Returns message 2.
Get a message from the queue. 
Returns message 3. 
Etc until message 5, then it starts from 1.

Expectation would be to get the message 1 for each basic-get as long as it's not acknowledged. This happens even if delivery-limit is set. 

We are preparing a test to show this behaviour as well. 

Thank you,
Anh

**Comment by michaelklishin** [comment]
Quorum queues starting with 3.10 have [two ways of requeueing](https://rabbitmq.com/quorum-queues.html#repeated-requeues) by design: to avoid unbound disk log growth with pathological consumers that consume and never acknowledge.

`basic.get` via the UI is not how you should be assessing message ordering. With concurrent consumers and requeueing, message ordering becomes a moot point: the queue will try to put it back into the original location but this can be immediately negated by operations of/on other consumers. They avoid this by requeueing back to the end of the queue instead of the original position (e.g. the head).

**Reply by kjnilsson** [reply]
Yes correct the only way to really ensure ordering is to configure the queue to use single active consumer _and_ ensure there is a delivery-limit set either via a declare arg or a policy.

**Reply by kjnilsson** [reply]
See https://www.rabbitmq.com/quorum-queues.html#repeated-requeues

**Comment by anhanhnguyen** [comment]
Hi,

Yes, we did the test with delivery-limit set and single active consumer.

I think this behaviour is from a list keeping track of requeued messages `returns`

When getting messages, messages from the head of `returns` will be taken, and the message is placed back on "Nack" (or requeued when consumers are down) to the end of this list.

By this way, the messages will be different each `basic.get` and nack-requeue true (messages cycle in the `returns` list)

We made a simple test here:
https://github.com/rabbitmq/rabbitmq-server/compare/main...anhanhnguyen:rabbitmq-server:main

We would like to confirm whether this is the intended design as we still expect the messages to go back to the "original order".

Thank you and best regards,
Anh

**Reply by kjnilsson** [reply]
I'd say it is more of a side-effect of the implementation that a deliberate decision. 

We have to consider what ordering guarantees queues in RabbitMQ actually can and do provide. Formally, queues in RabbitMQ are unordered as they allow multiple consumers to consume at the same time and ack and/or return messages in any order they like. 

RabbitMQ does provide slightly stronger guarantees than that though in that any messages sent from a single publishing channel must arrive at the queue in the order they were sent. However once these messages are consumed they effectively enter an unordered state and the quorum queue implementation doesn't violate that. instead any returned messages (assuming a delivery limit is in place) are ordered in the order they were returned. 

if you use single active consumer with quorum queues you can ensure a single consumer at a time receives messages and that they receive those messages in the input order of the queue. When SAC is configured basic.get is disallowed as that would violate the single active consumer restriction. There is one more factor your consumers need to ensure in that they don't reduce the prefetch used as this could theoretically also change the order of returned messages.

Ordering in messaging is a tricky subject and if your systems aren't able to cope with both disorder and duplication you may have a bad time using at-least-once queues. Streams may be a better option there but of course they put more complexity on the consuming applications in other ways.

**Reply by olikasg** [reply]
According to the broker semantics, even returns should maintain the message order in the queue in case there is a single consumer.
https://www.rabbitmq.com/semantics.html#ordering
> Messages can be returned to the queue using AMQP methods that feature a requeue parameter (`basic.recover`, `basic.reject` and  `basic.nack`), or due to a channel closing while holding unacknowledged messages. Any of these scenarios caused messages to be requeued at the back of the queue for RabbitMQ releases earlier than 2.7.0. **From RabbitMQ release 2.7.0, messages are always held in the queue in publication order, even in the presence of requeueing or channel closure.**

The same section states that the queueing order required by the AMQP spec is maintained which is still true. But if requeueing no longer maintains the publication order, it should be stated clearly and marked as a breaking change from previous semantics.

Whether relying on RabbitMQ for message ordering is good idea or not is a different subject.

**Reply by rjnilsson** [reply]
I'm part of the team that originally noticed this issue when switching from classic mirrored queues to quorum queues. A couple of things for background and clarification:

* The main issue here is that there seems to be a breaking change that is not documented. I've read and re-read the docs to try to ensure that we didn't misunderstand something regarding this.
* We don't really depend on the delivery order of the messages, at least not within a "safe" time window. We need to _process_ the messages in order, but we are maintaining a sorting/ordering buffer within the listening service which takes care of that, and we will not be running parallel instances of the service (the performance is good enough when running a single instance per queue).
* We detected this issue when the service got caught up in constant restarts due to logic errors in the incoming data (which we cannot affect), and we tried to remove the offending message(s) using the web admin interface (Reject requeue false), which _was_ working consistently when using classic queues. These messages were always at the head of the queue, and we could remove them safely before.
* We have set up delivery-limit as well as max-length (policy based) for the queues.

Yes, we could perhaps use a DLX, but we are in development and prefer to notice/detect this kind of logic errors in this way for the time being. But whether we should drop such messages depends on context as well.

**Reply by kjnilsson** [reply]
The behaviour when returned messages are returned at the back of the queue when no delivery-limit is set is documented here: https://www.rabbitmq.com/quorum-queues.html#repeated-requeues

I will add section to the quorum queue docs about return ordering in general.

**Reply by rjnilsson** [reply]
> The behaviour when returned messages are returned at the back of the queue when no delivery-limit is set is documented here: https://www.rabbitmq.com/quorum-queues.html#repeated-requeues

Yes. We are using that (don't know if I misunderstood your comment):

`We have set up delivery-limit as well as max-length (policy based) for the queues.`

With the "breaking change" I was only referring to the return ordering of quorum queues, in the presence of a single reader.

**Reply by kjnilsson** [reply]
You mean if the single reader returns messages in an order different to the one they were received they would be ordered in the order the were returned?

**Reply by rjnilsson** [reply]
That was pretty hard to parse :)

I don't know about that use case. I forgot to mention that we don't actually requeue explicitly; when we detect logic errors we simply bail out, the service dies, and the container is restarted. So just dropping the connection and expecting the original order again when reconnecting.

**Reply by kjnilsson** [reply]
For a single consumer that crashes any pending messages will be returned in the order they were assigned to the consumer. Are you seeing something different here?

**Reply by kjnilsson** [reply]
So we have a queue with `[A, B, C, D, E]` ready messages. The queue has a delivery-limit set and is configured to use single active consumer. A consumer attaches with a prefetch of 3 and gets assigned `[A, B, C]`. The consumer acks `A` then crashes. B and C gets added to the internal return queue in the order they were received, i.e. `[B, C]`. Another consumer attaches and gets assigned `[B, C, D]` as expected. That is how it should work. Do you see something different in your tests?

**Reply by rjnilsson** [reply]
Yes, our observation is different. I'll try with a more specific example:

* Quorum queue with delivery-limit (and max-length) with three nodes (IIRC)
* Single consumer
* Prefetch of 3 (this is not the actual prefetch value that we've been using)

Initial scenario:

1. Consumer connects to queue
2. Messages arrive [A, B, C, D, E]
3. Consumer receives [A, B, C] into prefetch (of course, this happens concurrently with 2 above)
4. Consumer processes and acks A, but crashes when attempting to process B
5. Consumer starts up, connects to queue
6. Consumer receives [B, C, D] into prefetch (actual order of reception not verified, but we order them internally)
7. Consumer crashes when attempting to process B
8. Return to 5 above

The above continues for some undefined time. We then turn off consumer auto-restart, and:

1. Start the web admin interface
2. Open the queue
3. Get single message from the queue (nack requeue true) and seemingly receives one random message of the set [B, C, D]
4. Get single message once again (nack requeue true) and seemingly receives some other random message of the set [B, C, D]
5. Etc

The "randomness" above are only illustrations (sometimes we actually see B first, but then something else), but I fail to see any pattern. We would expect to consistently receive B each time, as was our observation when using classic mirrored queues.

I also think that the example from @anhanhnguyen above shows this in a similar way.

Hope this helps.

**Comment by kjnilsson** [comment]
In a quorum queue issuing a basic.get is effectively implemented as a
consume with a prefetch of one. This is why basic.get is disallowed when
using single active consumer.

On Sun, 22 Jan 2023 at 17:26, rjnilsson ***@***.***> wrote:

> Yes, our observation is different. I'll try with a more specific example:
>
>    - Quorum queue with delivery-limit (and max-length) with three nodes
>    (IIRC)
>    - Single consumer
>    - Prefetch of 3 (this is not the actual prefetch value that we've been
>    using)
>
> Initial scenario:
>
>    1. Consumer connects to queue
>    2. Messages arrive [A, B, C, D, E]
>    3. Consumer receives [A, B, C] into prefetch (of course, this happens
>    concurrently with 2 above)
>    4. Consumer processes and acks A, but crashes when attempting to
>    process B
>    5. Consumer starts up, connects to queue
>    6. Consumer receives [B, C, D] into prefetch (actual order of
>    reception not verified, but we order them internally)
>    7. Consumer crashes when attempting to process B
>    8. Return to 5 above
>
> The above continues for some undefined time. We then turn off consumer
> auto-restart, and:
>
>    1. Start the web admin interface
>    2. Open the queue
>    3. Get single message from the queue (nack requeue true) and seemingly
>    receives one random message of the set [B, C, D]
>    4. Get single message once again (nack requeue true) and seemingly
>    receives some other random message of the set [B, C, D]
>    5. Etc
>
> The "randomness" above are only illustrations (sometimes we actually see B
> first, but then something else), but I fail to see any pattern. We would
> expect to consistently receive B each time, as was our observation when
> using classic mirrored queues.
>
> I also think that the example from @anhanhnguyen
> <https://github.com/anhanhnguyen> above shows this in a similar way.
>
> Hope this helps.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rabbitmq/rabbitmq-server/discussions/6805#discussioncomment-4750529>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJAHFBHDLWWLWXBS6TDIHLWTVUTJANCNFSM6AAAAAATR3YRXM>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***
> com>
>


-- 
*Karl Nilsson*

**Reply by rjnilsson** [reply]
Karl, thanks for your reply (again).

I'm neither an AMQP nor RabbitMQ expert, so for my understanding, would you care to explain to me how that applies to the described scenario and how the implementation leads to the observed behavior (or just point me to such information)?

I've been searching for a description of this behavior but haven't found anything regarding basic.get, quorum queues and the admin interface, except for the cases where SAC is applied, such as here: https://www.erlang-solutions.com/blog/rabbitmq-quorum-queues-explained-what-you-need-to-know/

**Reply by kjnilsson** [reply]
Getting a message from a queue and returning it is a mutating action, i.e. it will change the state of the queue just like attaching a consumer would and that is what you see here. I think a lot of our concerns will be alleviated by enabling single active consumer as this is a stricter mode w.r.t to return ordering as I that is what you are looking for. Your test above did not have single active consumer enabled. https://www.rabbitmq.com/consumers.html#single-active-consumer

Here is the right configuration to ensure consumers see messages in the order they were enqueued.

1. Create a quorum queue with single active consumer enabled.
2. Configure the queue with a delivery-limit
3. Ensure subsequent consumers do not have a lower prefetch than any prior consumer
4. Avoid explicitly returning individual messages for to be re-queued (why would you if you want to process in order?) (crashing is fine as it returns _all_ pending messages and the order is retained there).

I think there is an argument for making a change around point 3 to make sure what you can do that when single active consumer is enabled.

Yes we could consider _optionally_ doing continuous reordering of the returns queue in a future version of quorum queues. I see little benefit though as it isn't a guarantee you can program against without single active consumer and the risk is always to give an illusion of ordered processing when no such thing is technically possible when allowing for more than a single active consumer at a time. It would have a potentially noticeable performance penalty attached as well, especially when using many consumers with large prefetches but we'll consider it.

**Reply by rjnilsson** [reply]
I think we are still in a bit of a misunderstanding. Quorum queues works perfectly fine during normal usage for us. The ordering semantics are ok with respect to that.

This is related to attempting to peek at (and later remove) the head message from the queue using the admin interface without affecting the order of the currently existing messages, including the head message. This at a time when no subscribers except the admin interface interacts with, or are even attached to, the queue.

Peeking at the first message to see the contents ("nack + requeue true"), and if the contents is "bad" then remove it using "reject + requeue false", seems to be impossible with quorum queues. Kind of like the observer effect in physics, at least when judging by the externally observable behavior and disregarding implementation details.

We did not notice this behavior before switching to quorum queues. So, maybe my point at this time is simply that the behavior of getting a single messages from the queue, using the admin interface, when messages have been redelivered, seem to have changed. And that I can't find any documentation on that.

**Reply by mkuratczyk** [reply]
I've tried multiple times but was unable to reproduce. Perhaps I'm missing something here, this is what I tried:
```
# declare the queue
rabbitmqadmin declare queue name=qq queue_type=quorum arguments='{"x-delivery-limit": 5}'

# publish messages with payload from 'A' to 'Z'
for PAYLOAD in {A..Z}; do rabbitmqadmin publish routing_key=qq payload=$PAYLOAD; done

# consume with a prefetch of 3, wait 10 seconds before acking - this gives us time to kill the process
perf-test -x 0 -qq -u qq -qa x-delivery-limit=5 -q 3 -L 10000000

# kill the publisher after the first message is acked (this part is manual and time sensitive, of course could be replaced with a small app)
killall -9 java

# at this point I have 25 messages in the queue ('A' was acked, 'B' and 'C' requeued)
rabbitmqadmin get queue=qq ackmode=reject_requeue_true
+-------------+----------+---------------+---------+---------------+------------------+-------------+
| routing_key | exchange | message_count | payload | payload_bytes | payload_encoding | redelivered |
+-------------+----------+---------------+---------+---------------+------------------+-------------+
| qq          |          | 24            | B       | 1             | string           | True        |
+-------------+----------+---------------+---------+---------------+------------------+-------------+
```
(note, while the output says there are 24 messages in the queue, since the message is requeued, there are actually 25 again)

I also tried with the management UI but I always got 'B' as well (I believe they perform the same action under the hood).
If you can reproduce your issue using these steps or modify them accordingly, or write an app that demonstrates the problem, it'd be much easier to help/fix the issue.

**Reply by rjnilsson** [reply]
Thanks for helping out. I do have an active example of the behavior "in the wild" right now, that exposes the "different messages every get through the admin interface" but that is internally.

Did you see the sample reference earlier by @anhanhnguyen at https://github.com/rabbitmq/rabbitmq-server/compare/main...anhanhnguyen:rabbitmq-server:main

**Reply by mkuratczyk** [reply]
Thanks, I missed that. I'll leave it to Karl to decide whether this is a bug or a feature.

**Reply by rjnilsson** [reply]
Sounds great. Thanks to you both.
